src/modules/iotjs_module_mqtt.c

```diff
@@ -104,51 +96,67 @@ static void iotjs_mqtt_alloc_buf(unsigned char** buf, int expected_size,
 }
 
 JS_FUNCTION(MqttReadPacket) {
-  JS_DECLARE_THIS_PTR(mqtt, mqtt);
-  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_mqtt_t, mqtt);
-
   iotjs_bufferwrap_t* wrap = iotjs_bufferwrap_from_jbuffer(jargv[0]);
-  _this->src = (unsigned char*)iotjs_bufferwrap_buffer(wrap);
-  _this->srclen = (int)iotjs_bufferwrap_length(wrap);
-  _this->offset = 0;
-
-  unsigned char* buf = NULL;
-  int buf_size = 0;
-  iotjs_mqtt_alloc_buf(&buf, _this->srclen, &buf_size);
-  if (buf == NULL) {
-    return JS_CREATE_ERROR(COMMON, "mqtt payload buf create error on read");
-  }
-  int r = MQTTPacket_read(buf, buf_size, iotjs_mqtt_read_buf, _this);
-  if (r < 0) {
-    return JS_CREATE_ERROR(COMMON, "mqtt payload unknown header type");
-  }
+  unsigned char *buf = (unsigned char*)iotjs_bufferwrap_buffer(wrap);
+  int buf_size = (int)iotjs_bufferwrap_length(wrap);
 
-  jerry_value_t ret = jerry_create_object();
-  iotjs_jval_set_property_jval(ret, "type", jerry_create_number(r));
+  MQTTHeader header = {0};
+  header.byte = buf[0];
+  int type = header.bits.type;
 
-  if (r == CONNACK) {
-    unsigned char present;
-    unsigned char code;
-    if (MQTTDeserialize_connack(&present, &code, buf, buf_size) != 1 ||
-        code != 0) {
-      return JS_CREATE_ERROR(COMMON, "Unable to connect.");
-    }
-  } else if (r == SUBACK) {
-    int count = 0;
-    int grantedQoS = 0;
-    unsigned short msgId;
-    if (MQTTDeserialize_suback(&msgId, 1, &count, (int*)&grantedQoS, buf,
-                               buf_size) != 1 ||
-        grantedQoS == 0x80) {
-      return JS_CREATE_ERROR(COMMON, "SUBACK failed with server.");
+  jerry_value_t ret;
+  do {
+    if (type == CONNACK) {
+      unsigned char present;
+      unsigned char code;
+      int r = MQTTDeserialize_connack(&present, &code, buf, buf_size);
+      if (r !=1) {
+        ret = JS_CREATE_ERROR(COMMON, "CONNACK Decode Failed");
+        break;
+      }
+      if (code == 0x01) {
+        ret = JS_CREATE_ERROR(COMMON, "Unacceptable Protocol Version");
+        break;
+      }
+      if (code == 0x02) {
+        ret = JS_CREATE_ERROR(COMMON, "Identifier Rejected");
+        break;
+      }
+      if (code == 0x03) {
+        ret = JS_CREATE_ERROR(COMMON, "Server Unavailable");
+        break;
+      }
+      if (code == 0x04) {
+        ret = JS_CREATE_ERROR(COMMON, "Bad User Name Or Password");
+        break;
+      }
+      if (code == 0x05) {
+        ret = JS_CREATE_ERROR(COMMON, "Not Authorized");
+        break;
+      }
+      if (code != 0x00) {
+        char err_str[64];
+        sprintf(err_str, "Unknown CONNACK Code %d", code);
+        ret = JS_CREATE_ERROR(COMMON, err_str);
+        break;
+      }
```

Comment:
The code numbers seems to be some DEFINEs on libmqtt, could you please use those instead of bare numbers?